### PR TITLE
New base class for discrete/recompute stress materials

### LINKED ISSUE #7847: Base class for Update/compute=false materials
### Description of the enhancement or error report
Limiting the use of the discrete/recompute materials, set to be discrete by using `compute=false` in the parameters, to only a J2 radial return mapping method is too restrictive.  

### Rationale for the enhancement or information for reproducing the error
~~@bwspenc doesn't like the name Recompute and will take any excuse to change the name~~
Other materials, such as hot pressing or some damage models, could fit well as recompute materials but are not J2 radial return mapping materials.  A new base class in Tensor Mechanics, `UpdateStress`, for the recompute materials will be created.

### Identified impact
The base class change should be fairly seamless by having the existing `RecomputeRadialReturn` class inherit from the new base class.  Changing the names of the Recompute materials will require changing input files in MOOSE and in BISON.
@bwspenc @acasagran @novasr

### PR DESCRIPTION
**Pull request by sapitts**
### Relevant design information

Limiting the use of the recompute (`compute = false`) materials to just radial return mapping was too restrictive; this commit adds a new base class for the recompute materials, named `UpdateReturnMappingStressBase`.  The existing radial return material inherits from this new base class, and new non-radial return materials can also inherit from this base class.
### Test cases

~~No test cases are modified in this commit.~~ Updated the inheriting classes and the class names in the tests folders; no tests were regolded.

@bwspenc and/or @dschwen  please review so that @acasagran can use this class in his open pore compression model.

Closes #7847

### CODE REVIEW

File: modules/tensor_mechanics/include/materials/UpdateReturnMappingStressBase.h [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+
+//Forward declaration
+class UpdateReturnMappingStressBase;
+
+template<>
+InputParameters validParams<UpdateReturnMappingStressBase>();
+
+/**
+ * UpdateReturnMappingStressBase is a material that is not called by MOOSE because
+ * of the compute=false flag set in the parameter list.  This class is a base class
```
Comment (bwspenc) [inline modules/tensor_mechanics/include/materials/UpdateReturnMappingStressBase.h]: I'd reword the sentence: 
"This class is a base class for materials which perform internal Newton interations, such as radial return mapping materials."
to something like:
"This class is a base class for materials that use Newton iterations to perform radial return mapping iterations to compute inelastic strains."

**Review by bwspenc — COMMENTED** [review]
This all looks really good, but we need to work more on the names (I know, you're really surprised that I'd think that!).

Here are my issues:
1) `UpdateReturnMappingStressBase` is not accurately named because it is now general enough to apply to any model that computes inelastic strain, not just return mapping models.

2) When I suggested that we use the name 'Update' in this name, I was thinking that we would start to follow the convention that we are using in BISON where we put Update at the end (moving away from starting all of the model names with 'Compute').

**Comment by sapitts** [comment]
I'd originally thought that this commit would not require an application patch, but I was wrong; @permcody would you help me with the  application patch tomorrow please?

**Comment by permcody** [comment]
Of course :)

File: modules/tensor_mechanics/include/materials/StressUpdateBase.h [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+public:
+  StressUpdateBase(const InputParameters & parameters);
+
+  /// A radial return (J2) mapping method is defined in this material by overwritting
+  /// the computeInelasticStrainIncrement method.
+  virtual void updateStress(RankTwoTensor & strain_increment,
+                            RankTwoTensor & inelastic_strain_increment,
+                            RankTwoTensor & stress_new);
+
+  void setQp(unsigned qp);
```
Comment (permcody) [inline modules/tensor_mechanics/include/materials/StressUpdateBase.h]: I'd prefer to see `unsigned int` here.

File: modules/tensor_mechanics/src/materials/StressUpdateBase.C [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+
+void
+StressUpdateBase::updateStress(RankTwoTensor & strain_increment,
+                                     RankTwoTensor & inelastic_strain_increment,
+                                     RankTwoTensor & stress_new)
+{
+}
+
+void
+StressUpdateBase::setQp(unsigned qp)
```
Comment (permcody) [inline modules/tensor_mechanics/src/materials/StressUpdateBase.C]: `unsigned int` here too

**Review by permcody — APPROVED** [review]
This looks good, just two minor comments, but I'm willing to approve.

**Comment by sapitts** [comment]
Fixed the `unsigned int` and naming convention issues....and maybe got a little carried away with updating the inheriting classes, tests, and documentations (the XFEM documentation folder was actually a copy of everything in the UserObjects documentation folder).

**Comment by permcody** [comment]
Holy Dooley! I guess if @bwspenc signs off on this I'm good. Lots of new changes.

**Comment by sapitts** [comment]
It's a good test case for revoking approval at least?

**Comment by permcody** [comment]
Good point! I don't see how to revoke my approval. That's a problem.

**Review by permcody — DISMISSED** [review]
Test PR revoke.

**Comment by permcody** [comment]
Ah... That did it!

File: docs/content/modules/tensor_mechanics/introduction/Stresses.md [hunk 4]
```diff
@@ -95,15 +95,15 @@ The integration of the model is performed using a combination of Material and Di
...
+##Creep and Plasticity with StressUpdate Materials
+In addition to the User Object based plasticity models, another set of plasticity and creep material models have been developed with the `DiscreteMaterial` class, which allows for iterations within the material itself.  These iterative materials are designated by `StressUpdate` at the end of the class name. These classes use a Radial Return von Mises, or J2, approach to determine the inelastic, creep, damage, or plasticity strain increments at each time step.  The advantage of the recompute materials is the ability to combine multiple inelastic stress calculations, such as creep and plasticity. The [Stress Update Radial Return Mapping Algorithm](/Materials/RadialReturnStressUpdate.md) discusses the general algorithm to return the stress state to the yield surface.
```
Comment (bwspenc) [inline docs/content/modules/tensor_mechanics/introduction/Stresses.md]: `DiscreteMaterial` should be replaced with `StressUpdateBase`, and replace "recompute" with "stress update" in a few places.

File: docs/content/modules/tensor_mechanics/introduction/Stresses.md [hunk 5]
```diff
@@ -117,4 +117,4 @@ $$
 $$
 where G is the isotropic shear modulus, and $\sigma^{trial}_{effective}$ is the scalar von Mises trial stress.  
 
-When more than one radial recompute material is included in the simulation `ComputeReturnMappingStress` will iterate over the change in the calculated stress until the return stress has reached a stable value.  
+When more than one radial return stress update material is included in the simulation `ComputeReturnMappingStress` will iterate over the change in the calculated stress until the return stress has reached a stable value.
```
Comment (bwspenc) [inline docs/content/modules/tensor_mechanics/introduction/Stresses.md]: I think you should strike "radial return" in this sentence.

**Review by bwspenc — APPROVED** [review]
Just a couple of minor issues with the documentation as noted, but otherwise, I think this looks great. I love the new names for the "recompute" objects!

**Comment by bwspenc** [comment]
It should be noted that this is a departure from the naming convention currently used for the "plug and play" material models in tensor_mechanics. Currently, all of those models start with the name "Compute", which is something that I think we should move away from going forward.  

Having those names start with "Compute" is a way to define a sort of namespace to delineate these models from the legacy models in tensor_mechanics, but this is something that will be unnecessary as we deprecate and delete those legacy models, and something that I don't think we want to be encumbered with going forward. The name "Compute" doesn't do anything to help describe those models, as essentially every object in MOOSE could arguably have "Compute" thrown in front of it.

We decided on a naming convention for new TM-based objects in BISON, where we name them by concatenating the physical material we are modeling and the thing it is computing, in that order. For these "recompute" models, we put the "Update" keyword at the end. This PR brings the naming of the "recompute" models in tensor_mechanics in line with that syntax. It only affects the "recompute" models, which are still in flux and aren't being used much yet, so this shouldn't be very disruptive.

Down the road, I think we should start renaming the existing "Compute" models, but we should start naming new models using the new naming convention. In the new naming scheme, we'd basically just strike "Compute" from the front of the existing names. To keep the plug and play models separate from the other models, I'd suggest that we use subdirectories in the tensor_mechanics/src/materials directory.

Pinging @tonkmr and @dschwen for their input.

**Comment by sapitts** [comment]
I have an application patch, with @jiangwen84's blessing, ready for BISON

**Comment by sapitts** [comment]
@dschwen Would you give this new naming convention, changing the recompute materials to `StressUpdate` as we discussed in the naming convention meeting last week, a thumbs up or thumbs down with a new name suggestion today please?  @acasagran is waiting for this base class to finish his work in BISON.

File: modules/tensor_mechanics/include/materials/StressUpdateBase.h [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+ * iterations to compute inelastic strains.  All materials inheriting from this
+ * class must be called by a separate material, such as ComputeReturnMappingStress.
+ */
+class StressUpdateBase : public Material
+{
+public:
+  StressUpdateBase(const InputParameters & parameters);
+
+  /// A radial return (J2) mapping method is defined in this material by overwritting
+  /// the computeInelasticStrainIncrement method.
```
Comment (acasagran) [inline modules/tensor_mechanics/include/materials/StressUpdateBase.h]: Where is the computeInelasticStrainIncrement method? @sapitts

File: docs/content/modules/tensor_mechanics/systems/Materials/HyperbolicViscoplasticityStressUpdate.md [hunk 7]
```diff
@@ -1,10 +1,10 @@
...
-!description /Materials/RecomputeRadialReturnHyperbolicViscoplasticity
+#HyperbolicViscoplasticityStressUpdate
+!description /Materials/HyperbolicViscoplasticityStressUpdate
 
-!devel /Materials/RecomputeRadialReturnHyperbolicViscoplasticity float=right width=auto margin=20px padding=20px background-color=#F8F8F8
+!devel /Materials/HyperbolicViscoplasticityStressUpdate float=right width=auto margin=20px padding=20px background-color=#F8F8F8
 
 ## Description
-{!content/modules/tensor_mechanics/common_documentation/supplementalRecomputeRadialReturn.md!}
+{!content/modules/tensor_mechanics/common_documentation/supplementalRadialReturnStressUpdate.md!}
```
Comment (dschwen) [inline docs/content/modules/tensor_mechanics/systems/Materials/HyperbolicViscoplasticityStressUpdate.md]: This is including the page content? We could just link to the page, but I have no strong opinion on this.

File: modules/tensor_mechanics/include/materials/RadialReturnStressUpdate.h [hunk 8]
```diff
@@ -7,42 +7,38 @@
...
 {
 public:
-  RecomputeRadialReturn(const InputParameters & parameters);
+  RadialReturnStressUpdate(const InputParameters & parameters);
 
   /// A radial return (J2) mapping method is defined in this material by overwritting
   /// the computeInelasticStrainIncrement method.
-  virtual void computeStress(RankTwoTensor & strain_increment,
+  virtual void updateStress(RankTwoTensor & strain_increment,
                              RankTwoTensor & inelastic_strain_increment,
```
Comment (dschwen) [inline modules/tensor_mechanics/include/materials/RadialReturnStressUpdate.h]: indentation :-)

File: modules/tensor_mechanics/src/materials/ComputeReturnMappingStress.C [hunk 9]
```diff
@@ -6,17 +6,17 @@
...
 template<>
 InputParameters validParams<ComputeReturnMappingStress>()
 {
   InputParameters params = validParams<ComputeFiniteStrainElasticStress>();
   params.addClassDescription("Compute stress using a radial return mapping implementation for creep or creep combined with plasticity");
-  params.addParam<unsigned int>("max_iterations", 30, "Maximum number of newton iterations in the radial return material");
-  params.addParam<Real>("relative_tolerance", 1e-5, "Relative convergence tolerance for the newton iteration within the radial return material");
-  params.addParam<Real>("absolute_tolerance", 1e-5, "Absolute convergence tolerance for newton iteration within the radial return material");
-  params.addParam<bool>("output_iteration_info", false, "Set true to output newton iteration information from the radial return material");
+  params.addParam<unsigned int>("max_iterations", 30, "Maximum number of picard iterations over the stress change after all update materials are called");
```
Comment (dschwen) [inline modules/tensor_mechanics/src/materials/ComputeReturnMappingStress.C]: I guess this is technically correct, but we might want to reword this a bit to avoid any potential confusion with multiapp picard iterations.

File: modules/tensor_mechanics/src/materials/ComputeReturnMappingStress.C [hunk 10]
```diff
@@ -39,7 +39,7 @@ ComputeReturnMappingStress::initialSetup()
   std::vector<MaterialName> models = getParam<std::vector<MaterialName> >("return_mapping_models");
   for (unsigned i = 0; i < models.size(); ++i)
   {
-    MooseSharedPointer<RecomputeRadialReturn> rrr = MooseSharedNamespace::dynamic_pointer_cast<RecomputeRadialReturn>(getMaterialSharedPointerByName(models[i]));
+    MooseSharedPointer<StressUpdateBase> rrr = MooseSharedNamespace::dynamic_pointer_cast<StressUpdateBase>(getMaterialSharedPointerByName(models[i]));
```
Comment (dschwen) [inline modules/tensor_mechanics/src/materials/ComputeReturnMappingStress.C]: @permcody can you give some input here. Isn't the lifetime management of the material objects the job of the warehouse classes? If so we should probably just use raw pointers here. The design does not require smart pointers, or does it? The classes just get deleted by the warehouse at the end of the simulation.

File: modules/tensor_mechanics/src/materials/HyperbolicViscoplasticityStressUpdate.C [hunk 11]
```diff
@@ -67,7 +67,7 @@ RecomputeRadialReturnHyperbolicViscoplasticity::computeStressInitialize(Real eff
 }
 
 Real
-RecomputeRadialReturnHyperbolicViscoplasticity::computeResidual(Real effectiveTrialStress, Real scalar)
+HyperbolicViscoplasticityStressUpdate::computeResidual(Real effectiveTrialStress, Real scalar)
 {
   Real residual = 0.;
```
Comment (dschwen) [inline modules/tensor_mechanics/src/materials/HyperbolicViscoplasticityStressUpdate.C]: For consistency I'd change this to `0.0`

File: modules/tensor_mechanics/src/materials/IsotropicPlasticityStressUpdate.C [hunk 12]
```diff
@@ -24,8 +24,8 @@ InputParameters validParams<RecomputeRadialReturnIsotropicPlasticity>()
 }
 
 
-RecomputeRadialReturnIsotropicPlasticity::RecomputeRadialReturnIsotropicPlasticity(const InputParameters & parameters) :
-    RecomputeRadialReturn(parameters),
+IsotropicPlasticityStressUpdate::IsotropicPlasticityStressUpdate(const InputParameters & parameters) :
+    RadialReturnStressUpdate(parameters),
     _yield_stress_function(isParamValid("yield_stress_function") ? &getFunction("yield_stress_function") : NULL),
     _yield_stress(getParam<Real>("yield_stress")),
```
Comment (dschwen) [inline modules/tensor_mechanics/src/materials/IsotropicPlasticityStressUpdate.C]: This pattern in the above two lines is being dropped everywhere else. Just have a _single_ parameter 
```
_yield_stress(getFunction("yield_stress") ),
```
This allows users to specify a single number as well.

File: modules/tensor_mechanics/src/materials/RadialReturnStressUpdate.C [hunk 13]
```diff
@@ -40,7 +33,7 @@ RecomputeRadialReturn::RecomputeRadialReturn(const InputParameters & parameters)
 }
 
 void
-RecomputeRadialReturn::computeStress(RankTwoTensor & strain_increment,
+RadialReturnStressUpdate::updateStress(RankTwoTensor & strain_increment,
                                      RankTwoTensor & inelastic_strain_increment,
```
Comment (dschwen) [inline modules/tensor_mechanics/src/materials/RadialReturnStressUpdate.C]: indentation

File: modules/tensor_mechanics/src/materials/StressUpdateBase.C [hunk 14]
```diff
@@ -0,0 +1,42 @@
...
+#include "MooseMesh.h"
+
+template<>
+InputParameters validParams<StressUpdateBase>()
+{
+  InputParameters params = validParams<Material>();
+  params.addClassDescription("Calculates the effective inelastic strain increment required to return the isotropic stress state to a J2 yield surface.  This class is intended to be a parent class for classes with specific constitutive models.");
+  params.addParam<std::string>("base_name", "Optional parameter that allows the user to define multiple mechanics material systems on the same block, i.e. for multiple phases");
+
+  params.addParam<bool>("compute", false, "The return stress increment classes are intended to be iterative materials");
```
Comment (dschwen) [inline modules/tensor_mechanics/src/materials/StressUpdateBase.C]: or `prams.addPrivateParam<bool>("compute", false);` which you used above.

File: modules/tensor_mechanics/src/materials/TempDepHardeningStressUpdate.C [hunk 15]
```diff
@@ -65,7 +65,7 @@ RecomputeRadialReturnTempDepHardening::RecomputeRadialReturnTempDepHardening(con
 }
 
 void
-RecomputeRadialReturnTempDepHardening::computeStressInitialize(Real effectiveTrialStress)
+TempDepHardeningStressUpdate::computeStressInitialize(Real effectiveTrialStress)
```
Comment (dschwen) [inline modules/tensor_mechanics/src/materials/TempDepHardeningStressUpdate.C]: Not a fan of `TempDep`. Why not `TemperatureDependent`? Or is it `TemporaryDeposit`?

File: modules/tensor_mechanics/tests/recompute_radial_return/tests [hunk 16]
```diff
@@ -21,53 +21,53 @@
...
     type = 'RunException'
     input = 'isotropic_plasticity_errors.i'
-    cli_args = 'Materials/isotropic_plasticity_recompute/hardening_function=harden_func
+    cli_args = 'Materials/isotropic_plasticity/hardening_function=harden_func
     Materials/elasticity_tensor/type=ComputeIsotropicElasticityTensor Materials/elasticity_tensor/youngs_modulus=2.1e5
     Materials/elasticity_tensor/poissons_ratio=0.
-    Materials/isotropic_plasticity_recompute/yield_stress=-5
-    Materials/isotropic_plasticity_recompute/compute=false'
+    Materials/isotropic_plasticity/yield_stress=-5
+    Materials/isotropic_plasticity/compute=false'
```
Comment (dschwen) [inline modules/tensor_mechanics/tests/recompute_radial_return/tests]: is that rest ever run with `compute = true`? why? Also please indent for readability.

**Review by dschwen — CHANGES_REQUESTED** [review]

File: docs/content/modules/tensor_mechanics/systems/Materials/HyperbolicViscoplasticityStressUpdate.md (hunk 7 above)
Comment (sapitts) [inline docs/content/modules/tensor_mechanics/systems/Materials/HyperbolicViscoplasticityStressUpdate.md]: This allows me to duplicate common documentation on several pages without writing the same information multiple times; in this case it's a brief intro to J2 radial return and a picture of a yield surface - I like the option to have this information on the same page rather than a link away

File: modules/tensor_mechanics/src/materials/StressUpdateBase.C (hunk 14 above)
Comment (sapitts) [inline modules/tensor_mechanics/src/materials/StressUpdateBase.C]: set or suppress? or both?

File: modules/tensor_mechanics/src/materials/IsotropicPlasticityStressUpdate.C (hunk 12 above)
Comment (sapitts) [inline modules/tensor_mechanics/src/materials/IsotropicPlasticityStressUpdate.C]: Keeping the yield function and the hardening function below as pointers is necessary because the user can only specify one of the two (yield function or hardening function).  If there's another way to maintain this functionality with the simpler params command I'd be glad to switch

File: modules/tensor_mechanics/src/materials/StressUpdateBase.C (hunk 14 above)
Comment (dschwen) [inline modules/tensor_mechanics/src/materials/StressUpdateBase.C]: Well, if the parameter is already added in `Material` then set+suppress would be the right thing here.

File: modules/tensor_mechanics/src/materials/IsotropicPlasticityStressUpdate.C (hunk 12 above)
Comment (dschwen) [inline modules/tensor_mechanics/src/materials/IsotropicPlasticityStressUpdate.C]: What? No. Right now the user can only specify a *_constant or a *_function. _Both_ for hardening _and_ yield. What I'm saying is that a constant is a simple form of a function! Just have one parameter and make it a function. The user can still put `yield_stress = 3` or `yield_stress = my_awesome_function` in their input files!

Comment (sapitts) [inline modules/tensor_mechanics/src/materials/IsotropicPlasticityStressUpdate.C]: Oh, I see.  done.

Comment (sapitts) [inline modules/tensor_mechanics/src/materials/IsotropicPlasticityStressUpdate.C]: Turns out this change is not trivial to implement because of the inheriting classes and the use of timeDerivative on the hardening function.  I'd like to call scope creep on this change and address the function/ constant problem in a separate PR if that's acceptable.

Comment (bwspenc) [inline modules/tensor_mechanics/src/materials/IsotropicPlasticityStressUpdate.C]: I agree, @sapitts.  Although I agree with @dschwen 's suggestion, this doesn't really have anything to do with the changes in review here. We have multiple development efforts waiting on this PR, and need to get this in without expanding its scope.

**Comment by sapitts** [comment]
Comments address, except the smart pointer question, and I have a patch for Bison ready.

File: modules/tensor_mechanics/src/materials/ComputeReturnMappingStress.C (hunk 10 above)
Comment (sapitts) [inline modules/tensor_mechanics/src/materials/ComputeReturnMappingStress.C]: @permcody or @dschwen would you help me understand better how I should implement this change to raw pointers?  Is something like
`  std::string * rrr = {actually I have no idea where to look for what would replace the dynamic case}<StressUpdateBase>(getMaterialByName(models[i]))` sort of on the right track, or does the dynamic pointer cast remain with a raw pointer?

File: modules/tensor_mechanics/src/materials/ComputeReturnMappingStress.C [hunk 17]
```diff
@@ -37,9 +37,9 @@ void
 ComputeReturnMappingStress::initialSetup()
 {
   std::vector<MaterialName> models = getParam<std::vector<MaterialName> >("return_mapping_models");
-  for (unsigned i = 0; i < models.size(); ++i)
+  for (unsigned int i = 0; i < models.size(); ++i)
   {
-    MooseSharedPointer<RecomputeRadialReturn> rrr = MooseSharedNamespace::dynamic_pointer_cast<RecomputeRadialReturn>(getMaterialSharedPointerByName(models[i]));
+    StressUpdateBase * rrr = dynamic_cast<StressUpdateBase *>(&getMaterialByName(models[i]));
     if (rrr)
```
Comment (sapitts) [inline modules/tensor_mechanics/src/materials/ComputeReturnMappingStress.C]: @permcody @dschwen I'm not sure at all that this approach is the correct way to move to raw/dump pointers, but it seems to compile and pass the tests on my machine.  Please let me know if there's a better way to implement the pointers

Comment (permcody) [inline modules/tensor_mechanics/src/materials/ComputeReturnMappingStress.C]: That works, but there's no reason to turn a valid reference into a pointer. Just leave it as a reference intead:

`StressUpdateBase & rrr = dynamic_cast<StressUpdateBase &>(getMaterialByName(models[i]));`

Comment (sapitts) [inline modules/tensor_mechanics/src/materials/ComputeReturnMappingStress.C]: Is there a good way to change the reference to a pointer/ boolean so that the if statement will still work?

Comment (dschwen) [inline modules/tensor_mechanics/src/materials/ComputeReturnMappingStress.C]: Uhm... wait, how do you check if a `dynamic_cast` was successful if you use a reference? This technically should not be legal.

Comment (permcody) [inline modules/tensor_mechanics/src/materials/ComputeReturnMappingStress.C]: Right - sorry Stephanie. If the dynamic_cast _may_ fail you can't use a reference. Keep the pointer and do the check.

Comment (sapitts) [inline modules/tensor_mechanics/src/materials/ComputeReturnMappingStress.C]: oh phew, one thing done correctly today.  Thank you both for the help.  With any luck this will be ready to merge tomorrow morning.

**Comment by sapitts** [comment]
Comments addressed, and I have a patch for bison ready but will need some guidance on going through the application patch process.

**Review by dschwen — APPROVED** [review]
